### PR TITLE
Backport `fn {d,r}av1d_cdf_thread_update` code size reduction from dav1d 1.4.2

### PR DESCRIPTION
**Pull request by fbossen**
Reorder CDF arrays so that copying can be performed in larger blocks.

### CODE REVIEW

File: src/cdf.rs [hunk 1]
```diff
@@ -89,10 +90,9 @@ pub struct CdfCoefContext {
 
 #[derive(Clone)]
 #[repr(C)]
+// Buffers padded to [4]/[8]/[16] for SIMD where needed.
```
Comment (kkysen) [inline src/cdf.rs:93]:
```suggestion
/// Buffers padded to [4]/[8]/[16] for SIMD where needed.
#[derive(Clone)]
#[repr(C)]
```

File: src/cdf.rs [hunk 2]
```diff
@@ -28,10 +28,11 @@ use strum::EnumCount;
 #[derive(Clone, Default)]
 #[repr(C)]
 pub struct CdfContext {
-    pub m: CdfModeContext,
-    pub kfym: Align32<[[[u16; N_INTRA_PRED_MODES + 3]; 5]; 5]>,
     pub coef: CdfCoefContext,
+    pub m: CdfModeContext,
+    pub mi: CdfModeInterContext,
```
Comment (kkysen) [inline src/cdf.rs:33]: Where is this field coming from?  I don't see it in the C changes.

Comment (fbossen) [reply src/cdf.rs:33]: I basically split `struct CdfModeContext` into two structures such as to make it easier to copy them independently. The C code uses `offsetof()` to copy parts of `struct CdfModeContext`

Comment (kkysen) [reply src/cdf.rs:33]: Ah, okay, that makes sense.

File: src/cdf.rs [hunk 3]
```diff
@@ -4958,136 +4977,112 @@ pub(crate) fn rav1d_cdf_thread_update(
...
-    update_cdf_2d!(8, 6, m.angle_delta);
-    for k in 0..N_TX_SIZES - 1 {
-        update_cdf_2d!(3, cmp::min(k + 1, 2), m.txsz[k]);
-    }
-    update_cdf_3d!(2, N_INTRA_PRED_MODES, 6, m.txtp_intra1);
-    update_cdf_3d!(3, N_INTRA_PRED_MODES, 4, m.txtp_intra2);
-    update_bit_1d!(3, m.skip);
     for k in 0..BlockLevel::COUNT {
         update_cdf_2d!(4, dav1d_partition_type_count[k] as usize, m.partition[k]);
     }
```
Comment (kkysen) [inline src/cdf.rs:5006]: This is equivalent to this, right?

```c
    update_cdf_2d(4, N_PARTITIONS - 3, m.partition[BL_128X128]);
    for (int k = BL_64X64; k < BL_8X8; k++)
        update_cdf_2d(4, N_PARTITIONS - 1, m.partition[k]);
    update_cdf_2d(4, N_SUB8X8_PARTITIONS - 1, m.partition[BL_8X8]);
```

File: src/cdf.rs [hunk 4]
```diff
@@ -4958,136 +4977,112 @@ pub(crate) fn rav1d_cdf_thread_update(
...
-    update_bit_2d!(7, 3, m.pal_y);
-    update_bit_1d!(2, m.pal_uv);
+    update_cdf_2d!(2, 15, m.txtp_inter1);
+    update_cdf_1d!(11, m.txtp_inter2.0);
+    update_cdf_3d!(2, N_INTRA_PRED_MODES, 6, m.txtp_intra1);
+    update_cdf_3d!(3, N_INTRA_PRED_MODES, 4, m.txtp_intra2);
+    update_cdf_1d!(7, m.cfl_sign.0);
+    update_cdf_2d!(8, 6, m.angle_delta);
+    update_cdf_1d!(4, m.filter_intra.0);
+    update_cdf_2d!(3, (RAV1D_MAX_SEGMENTS - 1) as usize, m.seg_id);
```
Comment (kkysen) [inline src/cdf.rs:5015]:
```suggestion
    update_cdf_2d!(3, SegmentId::COUNT - 1, m.seg_id);
```

When rebased on `main`.

File: src/cdf.rs [hunk 5]
```diff
@@ -102,23 +102,44 @@ pub struct CdfModeContext {
...
+
+/// Buffers padded to [4]/[8]/[16] for SIMD where needed.
+#[derive(Clone)]
+#[repr(C)]
+pub struct CdfModeInterContext {
+    // inter/switch
+    pub y_mode: Align32<[[u16; N_INTRA_PRED_MODES + 3]; 4]>,
+    pub wedge_idx: Align32<[[u16; 16]; 9]>,
+    pub comp_inter_mode: Align16<[[u16; N_COMP_INTER_PRED_MODES]; 8]>,
+    pub filter: Align8<[[[u16; 4]; 8]; 2]>,
```
Comment (kkysen) [inline src/cdf.rs:139]:
```suggestion
    pub filter: Align8<[[[u16; Rav1dFilterMode::N_FILTERS]; 8]; 2]>,
```

**Review by kkysen — APPROVED** [review]
Just some little comments.  Sorry I took a while to review this.

File: src/cdf.rs (hunk 3 above)
Comment (fbossen) [reply src/cdf.rs:5006]: yes

**Review by kkysen — APPROVED** [review]

**Comment by kkysen** [comment]
@fbossen, can we merge this?

**Pull request merged**